clang/test/CIR/CodeGen/no-prototype.c

```diff
@@ -0,0 +1,51 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir-enable -emit-cir %s -o %t.cir
+// RUN: FileCheck --input-file=%t.cir %s
+
+// No-proto definition followed by a correct call.
+int noProto0(x) int x; { return x; }
+int test0(int x) {
+  // CHECK: cir.func @test0
+  return noProto0(x); // We know the definition. Should be a direct call.
+  // CHECK: %{{.+}} = cir.call @noProto0(%{{.+}})
+}
+
+// Declaration without prototype followed by its definition, then a correct call.
+//
+// Call to no-proto is made after definition, so a direct call can be used.
+int noProto1();
+int noProto1(int x) { return x; }
+int test1(int x) {
+  // CHECK: cir.func @test1
+  return noProto1(x);
+  // CHECK: %{{.+}} = cir.call @noProto1(%{{[0-9]+}})
+}
+
+// Declaration without prototype followed by a correct call, then its definition.
+//
+// Call to no-proto is made before definition, so a variadic call that takes anything
+// is created. Later, when the definition is found, no-proto is replaced.
+int noProto2();
+int test2(int x) {
+  return noProto2(x);
+  // CHECK: %[[#PTR:]] = cir.get_global @noProto2 : cir.ptr <!cir.func<!s32i (!s32i)>>
+  // CHECK: %[[#C_PTR:]] = cir.cast(bitcast, %[[#PTR]] : !cir.ptr<!cir.func<!s32i (!s32i)>>), !cir.ptr<!cir.func<!s32i (!s32i)>>
+  // CHECK: %{{.+}} = cir.call %[[#C_PTR]](%{{.+}}) : (!cir.ptr<!cir.func<!s32i (!s32i)>>, !s32i) -> !s32i
+}
+int noProto2(int x) { return x; }
```

Comment:
I've been thinking more about this case, and I wonder if we can handle this on-the-fly in a more simple way:
- First time you see `noProto2`, attach a `#cir.no_prototype` attribute to it.
- Emit the `cir.call` as if `noProto2` supported multiple arguments, the verifier should be able to look at the `#cir.no_prototype` and allow the mismatch. No casts or indirect calls needed.
- When we see the definition, we do the RAUW.

Does that make sense?

Reply:
I did think about something like that, but the idea of having a new attribute in CIR just for this hack bothers me a bit. Regardless, it would certainly be an improvement in the code gen.

Something that came to mind would be to have a variable enum attribute (`!cir.tag<no-proto>`) that could be used for any similar issue in the future. It seems more reusable than a dedicated `#cir.no_prototype`. WDYT?

Reply:
> I did think about something like that, but the idea of having a new attribute in CIR just for this hack bothers me a bit. Regardless, it would certainly be an improvement in the code gen.

The approach isn't hacky, the language has a dirty hack and we need to tag things as what they are hehe!

> Something that came to mind would be to have a variable enum attribute (!cir.tag<no-proto>) that could be used for any similar issue in the future. 

This comes back to the discussion with Hongtao on adding attributes, I don't know how this would compose well with other ones we might add, perhaps hard to justify until more of them come and we can properly categorize. My take would be to implement this as a optional UnitAttr directly on FuncOp, and tthe verifier requires it to be a declaration in order to be attached.

> It seems more reusable than a dedicated #cir.no_prototype. WDYT?

This attribute conveys semantics IMO.

Reply:
> The approach isn't hacky, the language has a dirty hack

:laughing: 

> hard to justify until more of them come and we can properly categorize

Fair. I'll refactor this to use the `#cir.no_prototype` approach then!